[IMPROVEMENT] Rounded background for mentions/highlights

artrmz opened the pull request:
@RocketChat/ios

Closes #633

![mentions-highlight](https://user-images.githubusercontent.com/3925166/33394135-af47fcde-d541-11e7-9e93-54ee03787979.png)

**To fix**
- [x] Add rounded background for mentions and highlights

- [x] Clean code

- [x] Adjust tests to background changes

- [x] Increase line spacing when there is mention/highlight in message

- [x] Fix background frame positioning for long messages

- [x] Align text to username's leading constraint

- [x] Fix content drawing after rotation from landscape to portrait

- [x] Fix text selection

- [x] lower the background for mentions to make it look more separated from each other

macabeus commented:
I liked your class `HighlightTextView`. It's useful for the pull request #823 in order to make `ChatMessageCell` more readable.
[I merged your branch with my branch](https://github.com/macabeus/Rocket.Chat.iOS/commit/977ee81fa056a5cf799a6ec2f0f4bf76de4d5772), to also use this class, then, when your pull request is merged, our features will be working together.
Thank you!

artrmz commented:
Added few changes. Also added increasing line spacing for whole message if it contains any highlight to make it look better (otherwise background was overlapping each other in some cases). Maybe we should add 10% of font size line spacing for all messages on the other hand?

rafaelks commented on Rocket.Chat/Views/Chat/HighlightTextView.swift:
```diff
@@ -0,0 +1,80 @@
+
+            for i in 0..<glyphRuns.count {
+                // swiftlint:disable force_cast
+                let run = glyphRuns[i] as! CTRun
```
Let's not disable any `force_cast`

artrmz replied:
👍

artrmz commented:
@rafaelks done as you suggested, although I am not sure if we can merge it until #878 is fixed anyway since it's kinda blocking my tests. Wanted to test it for very long msg with lots of mentions but right now can't do it (same issue on App Store version)

rafaelks commented:
@artrmz Looks great, thank you! Yes, let's fix this bug first before merging your PR.

artrmz commented:
Fixed #878 although gotta fix positioning a little bit more since it's a bit off for very long messages  (it starts to get off by few pixels after like 20th line of text)

artrmz commented:
Seems like it's all fixed now ;) Waiting for reviews. Let me know if should I resolve conflicts or you will do that while merging to develop.

@rafaelks @cardoso

rafaelks commented:
@artrmz Thank you for your awesome work! Just a minor thing... look at the screenshot, you see that the `mention` isn't aligned to the `username` on the top? Can you fix this, please?

![img_0150](https://user-images.githubusercontent.com/551004/32374483-0e747baa-c085-11e7-84f3-67dd57053af5.jpg)

artrmz commented:
Seems like easy thing to fix, will do that today or tomorrow :)

artrmz commented:
@rafaelks Fixed the aligmnent although found another issue ^^ It draws incorrectly (not always) after rotation.

artrmz commented:
Found a workaround for rotation issue although there is another issue with UITextView (probably on maste/develop branch as well) causing to resize incorrectly (wrong height) and cutting a text (last word of message) in some rare situations. Gotta investigate it further before I finish this one. I feel like writing my own UITextView (subclassing from UIView instead of UITextView) would be easier and give the better results in this situation 😱

cardoso commented:
@artrmz 

Text cutting can be fixed by tweaking this function in ChatMessageCell.swift:

```swift
    static func cellMediaHeightFor(message: Message, sequential: Bool = true) -> CGFloat {
        let fullWidth = UIScreen.main.bounds.size.width
        let attributedString = MessageTextCacheManager.shared.message(for: message)
        let height = attributedString?.heightForView(withWidth: fullWidth - 62)

        var total = (height ?? 0) + (sequential ? 8 : 29)
```

Tweaking some values there like adding 1 or 2 pixels to the height should fix it.

artrmz commented:
@cardoso Ah, thanks for explaining. That makes it a little bit more complicated. Although I think we shouldn't use some constants (62 a example here) like that. Should have taken those from view's constraints and subtract additional 16 for UITextView's scrollView insets if I am correct. That way we won't have to touch that method in case someones edits storyboard.

rafaelks commented:
@artrmz Yes! We may change the way we handle the size of the cell by just using AutoLayout, I'm not opposite to this idea, but will require a few changes in the code.

rafaelks commented:
@artrmz We changed the code a little bit, to process markdown in background & update on main-thread only. Do you mind fixing the conflicts? Maybe we could add this PR to `1.7.3` milestone. :-)

artrmz commented:
@rafaelks Yes, I'll solve all conflicts although it requires a bit more work until it's ready for merging.

rafaelks commented:
@artrmz This PR is looking really good, are we missing something to merge it? Really wanted to merge this new style. 👍

artrmz commented:
@rafaelks It breaks few things sadly and I did not have time to take care of it yet. For example text selection does not work.

cardoso commented:
@artrmz looks like we don't need selection anymore 😃 #1001

artrmz commented:
@cardoso Great :) Will take a look at this PR again since seems like it broke after all those changes on develop, probably something with mentions handling in the background.

artrmz commented:
I've refactored it completely. Should be ready to test "soon" :)

rafaelks commented:
@artrmz Thank you!!! 🎉

rafaelks commented:
@artrmz Thank you for your PR! I was testing here, and looks like the spacing between messages is too big. Do you also experience that?

![simulator screen shot - iphone se - 2017-11-30 at 09 49 42](https://user-images.githubusercontent.com/551004/33429402-dc73bf8c-d5b3-11e7-959c-7ed0b97c0d46.png)

artrmz commented:
@rafaelks Will check this out. Thanks!

artrmz commented:
@rafaelks Should be fixed now
![zrzut ekranu 2017-11-30 o 14 28 25](https://user-images.githubusercontent.com/3925166/33433147-c6459e48-d5da-11e7-8182-9c489daf9ad1.png)

artrmz commented:
Restored isSelectable as #1027 needed it.

rafaelks commented:
@artrmz This is working great to me know... may I ask you just one more thing? The left-edge and right-edge on a mention background, can it be a little bigger? Because the spacing right now between the background color and the font is almost 0px.

Example here:

![chat](https://user-images.githubusercontent.com/551004/33631478-d1d52e52-d9f1-11e7-8c0b-9d79d9480f6f.png)

artrmz commented:
@rafaelks The easiest way to achieve that would be to add nonbreaking spaces at start/end of the mention. Dunno if we want to do such thing? Or can try to add few pixels on each side, will give it a try tomorrow.

artrmz commented:
@rafaelks That's best I could achieve at the moment.

![zrzut ekranu 2017-12-06 o 09 02 05](https://user-images.githubusercontent.com/3925166/33650904-2728a290-da64-11e7-884c-a05fad010a01.png)
![zrzut ekranu 2017-12-06 o 09 02 02](https://user-images.githubusercontent.com/3925166/33650905-2754889c-da64-11e7-83fa-9a66984b1ac4.png)

I think it looks better without those changes ;) It will either clip to bounds and be cut if mention is the first word in the line OR can add padding to whole text but it won't be aligned with the username. It will be really hard to detect if first word in line is a mention and move the whole line.

rafaelks commented:
@artrmz Currently way looks better, but only because it's cutting on the left... let's merge this way now and try to improve it over the time! 👍. Thanks a lot for looking into it.
